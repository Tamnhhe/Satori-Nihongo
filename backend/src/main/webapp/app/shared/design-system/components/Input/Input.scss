@import '../../tokens/variables';

.satori-input-wrapper {
  display: flex;
  flex-direction: column;
  gap: $spacing-2;

  &--full-width {
    width: 100%;
  }

  &--disabled {
    opacity: 0.6;
    pointer-events: none;
  }

  &--loading {
    position: relative;

    &::after {
      content: '';
      position: absolute;
      top: 50%;
      right: $spacing-3;
      width: 16px;
      height: 16px;
      border: 2px solid $gray-300;
      border-top-color: $primary-600;
      border-radius: 50%;
      animation: spin 1s linear infinite;
      z-index: 1;
    }
  }
}

.satori-input-label {
  font-size: $text-sm;
  font-weight: $font-weight-medium;
  color: $gray-700;
  margin-bottom: $spacing-1;

  &__required {
    color: $error-500;
    margin-left: $spacing-1;
  }
}

.satori-input-container {
  position: relative;
  display: flex;
  align-items: center;
}

.satori-input {
  width: 100%;
  font-family: $font-family-sans;
  font-weight: $font-weight-normal;
  border-radius: $radius-md;
  border: 1px solid $gray-300;
  background-color: white;
  color: $gray-900;
  transition: all $transition-fast;

  &:focus {
    outline: none;
    border-color: $primary-500;
    box-shadow: 0 0 0 3px rgba($primary-500, 0.1);
  }

  &:hover:not(:disabled):not(:focus) {
    border-color: $gray-400;
  }

  &::placeholder {
    color: $gray-400;
  }

  &:disabled {
    background-color: $gray-50;
    color: $gray-500;
    cursor: not-allowed;
  }

  // Size variants
  .satori-input-wrapper--sm & {
    height: $input-height-sm;
    padding: 0 $spacing-3;
    font-size: $text-sm;
  }

  .satori-input-wrapper--md & {
    height: $input-height-md;
    padding: 0 $spacing-4;
    font-size: $text-base;
  }

  .satori-input-wrapper--lg & {
    height: $input-height-lg;
    padding: 0 $spacing-5;
    font-size: $text-lg;
  }

  // Variant styles
  .satori-input-wrapper--filled & {
    background-color: $gray-50;
    border-color: transparent;

    &:focus {
      background-color: white;
      border-color: $primary-500;
    }

    &:hover:not(:disabled):not(:focus) {
      background-color: $gray-100;
    }
  }

  // With icons
  &--with-left-icon {
    .satori-input-wrapper--sm & {
      padding-left: calc(#{$spacing-3} + 16px + #{$spacing-2});
    }

    .satori-input-wrapper--md & {
      padding-left: calc(#{$spacing-4} + 20px + #{$spacing-2});
    }

    .satori-input-wrapper--lg & {
      padding-left: calc(#{$spacing-5} + 24px + #{$spacing-2});
    }
  }

  &--with-right-icon {
    .satori-input-wrapper--sm & {
      padding-right: calc(#{$spacing-3} + 16px + #{$spacing-2});
    }

    .satori-input-wrapper--md & {
      padding-right: calc(#{$spacing-4} + 20px + #{$spacing-2});
    }

    .satori-input-wrapper--lg & {
      padding-right: calc(#{$spacing-5} + 24px + #{$spacing-2});
    }
  }

  // Error state
  .satori-input-wrapper--error & {
    border-color: $error-500;

    &:focus {
      border-color: $error-500;
      box-shadow: 0 0 0 3px rgba($error-500, 0.1);
    }
  }
}

.satori-textarea {
  min-height: 80px;
  padding: $spacing-3 $spacing-4;
  line-height: $leading-relaxed;

  &--resize-none {
    resize: none;
  }

  &--resize-vertical {
    resize: vertical;
  }

  &--resize-horizontal {
    resize: horizontal;
  }

  &--resize-both {
    resize: both;
  }

  .satori-input-wrapper--sm & {
    min-height: 60px;
    padding: $spacing-2 $spacing-3;
    font-size: $text-sm;
  }

  .satori-input-wrapper--lg & {
    min-height: 100px;
    padding: $spacing-4 $spacing-5;
    font-size: $text-lg;
  }
}

.satori-input-icon {
  position: absolute;
  display: flex;
  align-items: center;
  justify-content: center;
  color: $gray-400;
  pointer-events: none;

  &--left {
    .satori-input-wrapper--sm & {
      left: $spacing-3;
      width: 16px;
      height: 16px;
    }

    .satori-input-wrapper--md & {
      left: $spacing-4;
      width: 20px;
      height: 20px;
    }

    .satori-input-wrapper--lg & {
      left: $spacing-5;
      width: 24px;
      height: 24px;
    }
  }

  &--right {
    .satori-input-wrapper--sm & {
      right: $spacing-3;
      width: 16px;
      height: 16px;
    }

    .satori-input-wrapper--md & {
      right: $spacing-4;
      width: 20px;
      height: 20px;
    }

    .satori-input-wrapper--lg & {
      right: $spacing-5;
      width: 24px;
      height: 24px;
    }
  }

  &--clickable {
    pointer-events: auto;
    cursor: pointer;
    color: $gray-500;

    &:hover {
      color: $gray-700;
    }
  }

  .satori-input-wrapper--error & {
    color: $error-500;
  }
}

.satori-input-message {
  font-size: $text-sm;
  line-height: $leading-tight;

  &--error {
    color: $error-600;
  }

  &--helper {
    color: $gray-500;
  }
}

// Animations
@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
