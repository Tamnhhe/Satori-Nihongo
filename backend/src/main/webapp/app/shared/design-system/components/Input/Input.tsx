import React, { forwardRef, useId } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { useAriaAttributes } from '../../../hooks/useAccessibility';
import './Input.scss';

export interface InputProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {
  label?: string;
  error?: string;
  helperText?: string;
  size?: 'sm' | 'md' | 'lg';
  variant?: 'outline' | 'filled';
  leftIcon?: IconDefinition;
  rightIcon?: IconDefinition;
  onRightIconClick?: () => void;
  rightIconLabel?: string;
  fullWidth?: boolean;
  required?: boolean;
  loading?: boolean;
  showRequiredIndicator?: boolean;
}

export const Input = forwardRef<HTMLInputElement, InputProps>(
  (
    {
      label,
      error,
      helperText,
      size = 'md',
      variant = 'outline',
      leftIcon,
      rightIcon,
      onRightIconClick,
      rightIconLabel,
      fullWidth = false,
      required = false,
      loading = false,
      showRequiredIndicator = true,
      className = '',
      id,
      ...props
    },
    ref,
  ) => {
    const generatedId = useId();
    const inputId = id || generatedId;
    const hasError = !!error;
    const errorId = `${inputId}-error`;
    const helperId = `${inputId}-helper`;

    const wrapperClasses = [
      'satori-input-wrapper',
      `satori-input-wrapper--${size}`,
      `satori-input-wrapper--${variant}`,
      fullWidth && 'satori-input-wrapper--full-width',
      hasError && 'satori-input-wrapper--error',
      loading && 'satori-input-wrapper--loading',
      props.disabled && 'satori-input-wrapper--disabled',
      required && 'satori-input-wrapper--required',
      className,
    ]
      .filter(Boolean)
      .join(' ');

    const inputClasses = [
      'satori-input',
      leftIcon && 'satori-input--with-left-icon',
      rightIcon && 'satori-input--with-right-icon',
      'touch-target', // Ensure minimum touch target size
    ]
      .filter(Boolean)
      .join(' ');

    // Build describedBy string
    const describedByIds = [];
    if (error) describedByIds.push(errorId);
    if (helperText && !error) describedByIds.push(helperId);
    const describedBy = describedByIds.length > 0 ? describedByIds.join(' ') : undefined;

    // ARIA attributes
    const ariaAttributes = useAriaAttributes(
      undefined,
      undefined,
      undefined,
      describedBy,
      undefined,
      undefined,
      props.disabled,
      required,
      hasError,
    );

    return (
      <div className={wrapperClasses}>
        {label && (
          <label htmlFor={inputId} className="satori-input-label">
            {label}
            {required && showRequiredIndicator && (
              <span className="satori-input-label__required" aria-label="required">
                *
              </span>
            )}
          </label>
        )}

        <div className="satori-input-container">
          {leftIcon && (
            <div className="satori-input-icon satori-input-icon--left" aria-hidden="true">
              <FontAwesomeIcon icon={leftIcon} />
            </div>
          )}

          <input ref={ref} id={inputId} className={inputClasses} {...ariaAttributes} {...props} />

          {loading && (
            <div className="satori-input-icon satori-input-icon--right" aria-hidden="true">
              <FontAwesomeIcon icon={faSpinner} className="satori-input-icon--spinning" />
            </div>
          )}

          {rightIcon && !loading && (
            <div
              className={`satori-input-icon satori-input-icon--right ${onRightIconClick ? 'satori-input-icon--clickable' : ''}`}
              onClick={onRightIconClick}
              role={onRightIconClick ? 'button' : undefined}
              tabIndex={onRightIconClick ? 0 : undefined}
              aria-label={rightIconLabel}
              onKeyDown={
                onRightIconClick
                  ? e => {
                      if (e.key === 'Enter' || e.key === ' ') {
                        e.preventDefault();
                        onRightIconClick();
                      }
                    }
                  : undefined
              }
            >
              <FontAwesomeIcon icon={rightIcon} aria-hidden="true" />
            </div>
          )}
        </div>

        {error && (
          <div id={errorId} className="satori-input-message satori-input-message--error error-message" role="alert" aria-live="polite">
            {error}
          </div>
        )}

        {helperText && !error && (
          <div id={helperId} className="satori-input-message satori-input-message--helper">
            {helperText}
          </div>
        )}
      </div>
    );
  },
);

Input.displayName = 'Input';

export interface TextareaProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  label?: string;
  error?: string;
  helperText?: string;
  size?: 'sm' | 'md' | 'lg';
  variant?: 'outline' | 'filled';
  fullWidth?: boolean;
  required?: boolean;
  resize?: 'none' | 'vertical' | 'horizontal' | 'both';
}

export const Textarea = forwardRef<HTMLTextAreaElement, TextareaProps>(
  (
    {
      label,
      error,
      helperText,
      size = 'md',
      variant = 'outline',
      fullWidth = false,
      required = false,
      resize = 'vertical',
      className = '',
      id,
      ...props
    },
    ref,
  ) => {
    const textareaId = id || `textarea-${Math.random().toString(36).substr(2, 9)}`;
    const hasError = !!error;

    const wrapperClasses = [
      'satori-input-wrapper',
      'satori-input-wrapper--textarea',
      `satori-input-wrapper--${size}`,
      `satori-input-wrapper--${variant}`,
      fullWidth && 'satori-input-wrapper--full-width',
      hasError && 'satori-input-wrapper--error',
      props.disabled && 'satori-input-wrapper--disabled',
      className,
    ]
      .filter(Boolean)
      .join(' ');

    const textareaClasses = ['satori-input', 'satori-textarea', `satori-textarea--resize-${resize}`].filter(Boolean).join(' ');

    return (
      <div className={wrapperClasses}>
        {label && (
          <label htmlFor={textareaId} className="satori-input-label">
            {label}
            {required && <span className="satori-input-label__required">*</span>}
          </label>
        )}

        <div className="satori-input-container">
          <textarea
            ref={ref}
            id={textareaId}
            className={textareaClasses}
            aria-invalid={hasError}
            aria-describedby={error ? `${textareaId}-error` : helperText ? `${textareaId}-helper` : undefined}
            {...props}
          />
        </div>

        {error && (
          <div id={`${textareaId}-error`} className="satori-input-message satori-input-message--error">
            {error}
          </div>
        )}

        {helperText && !error && (
          <div id={`${textareaId}-helper`} className="satori-input-message satori-input-message--helper">
            {helperText}
          </div>
        )}
      </div>
    );
  },
);

Textarea.displayName = 'Textarea';
