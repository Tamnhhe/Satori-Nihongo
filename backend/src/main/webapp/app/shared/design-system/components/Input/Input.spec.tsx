import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';
import { Input } from './Input';

expect.extend(toHaveNoViolations);

describe('Input Component', () => {
  it('should render with default props', () => {
    render(<Input placeholder="Enter text" />);
    const input = screen.getByPlaceholderText('Enter text');
    expect(input).toBeInTheDocument();
    expect(input).toHaveClass('form-control');
  });

  it('should render with label', () => {
    render(<Input label="Username" placeholder="Enter username" />);

    expect(screen.getByLabelText('Username')).toBeInTheDocument();
    expect(screen.getByText('Username')).toBeInTheDocument();
  });

  it('should handle different input types', () => {
    const { rerender } = render(<Input type="email" />);
    expect(screen.getByRole('textbox')).toHaveAttribute('type', 'email');

    rerender(<Input type="password" />);
    expect(screen.getByLabelText(/password/i)).toHaveAttribute('type', 'password');

    rerender(<Input type="number" />);
    expect(screen.getByRole('spinbutton')).toHaveAttribute('type', 'number');
  });

  it('should handle value changes', async () => {
    const handleChange = jest.fn();
    const user = userEvent.setup();

    render(<Input onChange={handleChange} />);
    const input = screen.getByRole('textbox');

    await user.type(input, 'test value');
    expect(handleChange).toHaveBeenCalledTimes(10); // 'test value' has 10 characters
    expect(input).toHaveValue('test value');
  });

  it('should show validation states', () => {
    const { rerender } = render(<Input error="This field is required" placeholder="Error state" />);

    expect(screen.getByPlaceholderText('Error state')).toHaveClass('is-invalid');
    expect(screen.getByText('This field is required')).toBeInTheDocument();
    expect(screen.getByTestId('error-message')).toHaveClass('invalid-feedback');

    rerender(<Input success="Valid input" placeholder="Success state" />);

    expect(screen.getByPlaceholderText('Success state')).toHaveClass('is-valid');
    expect(screen.getByText('Valid input')).toBeInTheDocument();
  });

  it('should handle disabled state', () => {
    render(<Input disabled placeholder="Disabled input" />);
    const input = screen.getByPlaceholderText('Disabled input');
    expect(input).toBeDisabled();
    expect(input).toHaveClass('form-control-disabled');
  });

  it('should render with prefix and suffix', () => {
    render(<Input prefix="$" suffix=".00" placeholder="Amount" />);

    expect(screen.getByTestId('input-prefix')).toHaveTextContent('$');
    expect(screen.getByTestId('input-suffix')).toHaveTextContent('.00');
  });

  it('should render with icon', () => {
    render(<Input icon="search" placeholder="Search..." />);
    expect(screen.getByTestId('input-icon')).toBeInTheDocument();
  });

  it('should handle different sizes', () => {
    const { rerender } = render(<Input size="small" />);
    expect(screen.getByRole('textbox')).toHaveClass('form-control-sm');

    rerender(<Input size="large" />);
    expect(screen.getByRole('textbox')).toHaveClass('form-control-lg');
  });

  it('should show character count when maxLength is provided', async () => {
    const user = userEvent.setup();
    render(<Input maxLength={10} showCount />);

    const input = screen.getByRole('textbox');
    await user.type(input, 'test');

    expect(screen.getByTestId('character-count')).toHaveTextContent('4/10');
  });

  it('should handle focus and blur events', async () => {
    const handleFocus = jest.fn();
    const handleBlur = jest.fn();
    const user = userEvent.setup();

    render(<Input onFocus={handleFocus} onBlur={handleBlur} placeholder="Focus test" />);

    const input = screen.getByPlaceholderText('Focus test');

    await user.click(input);
    expect(handleFocus).toHaveBeenCalledTimes(1);

    await user.tab();
    expect(handleBlur).toHaveBeenCalledTimes(1);
  });

  it('should support keyboard navigation', async () => {
    const user = userEvent.setup();
    render(<Input placeholder="Keyboard test" />);

    const input = screen.getByPlaceholderText('Keyboard test');

    await user.tab();
    expect(input).toHaveFocus();

    await user.keyboard('Hello World');
    expect(input).toHaveValue('Hello World');
  });

  it('should be accessible', async () => {
    const { container } = render(<Input label="Accessible Input" placeholder="Enter value" aria-describedby="help-text" required />);

    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should have proper ARIA attributes', () => {
    render(<Input label="Required Field" required error="This field is required" aria-describedby="error-message" />);

    const input = screen.getByLabelText('Required Field');
    expect(input).toHaveAttribute('required');
    expect(input).toHaveAttribute('aria-invalid', 'true');
    expect(input).toHaveAttribute('aria-describedby');
  });

  it('should clear input when clear button is clicked', async () => {
    const handleChange = jest.fn();
    const user = userEvent.setup();

    render(<Input clearable onChange={handleChange} defaultValue="test" />);

    const clearButton = screen.getByTestId('clear-button');
    await user.click(clearButton);

    expect(handleChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: expect.objectContaining({ value: '' }),
      }),
    );
  });

  it('should match snapshot', () => {
    const { container } = render(
      <div>
        <Input label="Basic Input" placeholder="Basic" />
        <Input label="Advanced Input" type="email" prefix="@" suffix=".com" error="Invalid email" required />
        <Input disabled placeholder="Disabled" />
        <Input loading placeholder="Loading" />
      </div>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
