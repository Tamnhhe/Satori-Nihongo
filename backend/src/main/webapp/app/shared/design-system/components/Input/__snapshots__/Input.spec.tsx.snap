// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should match snapshot 1`] = `
<div>
  <div
    class="satori-input-wrapper satori-input-wrapper--md satori-input-wrapper--outline"
  >
    <label
      class="satori-input-label"
      for=":rf:"
    >
      Basic Input
    </label>
    <div
      class="satori-input-container"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="satori-input touch-target"
        id=":rf:"
        placeholder="Basic"
      />
    </div>
  </div>
  <div
    class="satori-input-wrapper satori-input-wrapper--md satori-input-wrapper--outline satori-input-wrapper--error satori-input-wrapper--required"
  >
    <label
      class="satori-input-label"
      for=":rg:"
    >
      Advanced Input
      <span
        aria-label="required"
        class="satori-input-label__required"
      >
        *
      </span>
    </label>
    <div
      class="satori-input-container"
    >
      <input
        aria-describedby=":rg:-error"
        aria-invalid="true"
        aria-required="true"
        class="satori-input touch-target"
        id=":rg:"
        prefix="@"
        suffix=".com"
        type="email"
      />
    </div>
    <div
      aria-live="polite"
      class="satori-input-message satori-input-message--error error-message"
      id=":rg:-error"
      role="alert"
    >
      Invalid email
    </div>
  </div>
  <div
    class="satori-input-wrapper satori-input-wrapper--md satori-input-wrapper--outline satori-input-wrapper--disabled"
  >
    <div
      class="satori-input-container"
    >
      <input
        aria-disabled="true"
        aria-invalid="false"
        aria-required="false"
        class="satori-input touch-target"
        disabled=""
        id=":rh:"
        placeholder="Disabled"
      />
    </div>
  </div>
  <div
    class="satori-input-wrapper satori-input-wrapper--md satori-input-wrapper--outline satori-input-wrapper--loading"
  >
    <div
      class="satori-input-container"
    >
      <input
        aria-invalid="false"
        aria-required="false"
        class="satori-input touch-target"
        id=":ri:"
        placeholder="Loading"
      />
      <div
        aria-hidden="true"
        class="satori-input-icon satori-input-icon--right"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-spinner satori-input-icon--spinning"
          data-icon="spinner"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M304 48a48 48 0 1 0 -96 0 48 48 0 1 0 96 0zm0 416a48 48 0 1 0 -96 0 48 48 0 1 0 96 0zM48 304a48 48 0 1 0 0-96 48 48 0 1 0 0 96zm464-48a48 48 0 1 0 -96 0 48 48 0 1 0 96 0zM142.9 437A48 48 0 1 0 75 369.1 48 48 0 1 0 142.9 437zm0-294.2A48 48 0 1 0 75 75a48 48 0 1 0 67.9 67.9zM369.1 437A48 48 0 1 0 437 369.1 48 48 0 1 0 369.1 437z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
